File: 768_max_chunks_to_sorted/max_chunks_to_sorted.go
/*
 * @Date: 2022-08-15
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-08-15
 * @FilePath: /algorithm/768_max_chunks_to_sorted/max_chunks_to_sorted.go
 */

package main

func maxChunksToSorted(arr []int) int {
	st := []int{}
	for _, x := range arr {
		if len(st) == 0 || x >= st[len(st)-1] {
			st = append(st, x)
		} else {
			mx := st[len(st)-1]
			st = st[:len(st)-1]
			for len(st) > 0 && st[len(st)-1] > x {
				st = st[:len(st)-1]
			}
			st = append(st, mx)
		}
	}
	return len(st)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(maxChunksToSorted([]int{5, 4, 3, 2, 1}) == 1)
	assert(maxChunksToSorted([]int{2, 1, 3, 4, 4}) == 4)
}

File: rust/src/769_max_chunks_to_sorted/max_chunks_to_sorted.rs
/*
 * @Date: 2022-10-13
 * @LastEditors: mengsen_wang@163.com
 * @LastEditTime: 2022-10-13
 * @FilePath: /algorithm/769_max_chunks_to_sorted/max_chunks_to_sorted.rs
 */

pub fn max_chunks_to_sorted(arr: Vec<i32>) -> i32 {
    let (mut m, mut res) = (0, 0);
    for i in 0..arr.len() {
        m = m.max(arr[i]);
        if m == i as i32 {
            res += 1;
        }
    }
    res
}

fn main() {
    {
        let arr = vec![4, 3, 2, 1, 0];
        let ans = 1;
        assert_eq!(max_chunks_to_sorted(arr), ans);
    }

    {
        let arr = vec![1, 0, 2, 3, 4];
        let ans = 4;
        assert_eq!(max_chunks_to_sorted(arr), ans);
    }
}
